processors/enduserprocessor/enduserprocessor.go

```diff
@@ -0,0 +1,534 @@
+package enduserprocessor
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/PaesslerAG/jsonpath"
+	"github.com/dgrijalva/jwt-go"
+	jsoniter "github.com/json-iterator/go"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/processor/processorhelper"
+	"go.uber.org/zap"
+
+	"github.com/hypertrace/collector/processors/enduserprocessor/hash"
+	"github.com/hypertrace/collector/processors/piifilterprocessor"
+)
+
+const (
+	enduserIDAttribute      = "enduser.id"
+	enduserRoleAttribute    = "enduser.role"
+	enduserScopeAttribute   = "enduser.scope"
+	enduserSessionAttribute = "session.id"
+)
+
+type config struct {
+	EndUser
+	hashAlgorithm hash.Algorithm
+}
+
+type processor struct {
+	logger             *zap.Logger
+	attributeConfigMap map[string][]config
+}
+
+type user struct {
+	id      string
+	role    string
+	scope   string
+	session string
+}
+
+var _ processorhelper.TProcessor = (*processor)(nil)
+
+func (p *processor) ProcessTraces(_ context.Context, traces pdata.Traces) (pdata.Traces, error) {
+	rss := traces.ResourceSpans()
+	for i := 0; i < rss.Len(); i++ {
+		rs := rss.At(i)
+
+		ilss := rs.InstrumentationLibrarySpans()
+		for j := 0; j < ilss.Len(); j++ {
+			ils := ilss.At(j)
+
+			spans := ils.Spans()
+			for k := 0; k < spans.Len(); k++ {
+				span := spans.At(k)
+
+				span.Attributes().ForEach(func(k string, v pdata.AttributeValue) {
+					unindexedKey := piifilterprocessor.UnindexedKey(k)
+					if configs, ok := p.attributeConfigMap[unindexedKey]; ok {
+						for _, cfg := range configs {
+							p.capture(span, cfg, v.StringVal())
+						}
+					}
+				})
+			}
+		}
+	}
+
+	return traces, nil
+}
+
+func (p *processor) capture(span pdata.Span, enduser config, value string) {
+	if !p.passesConditions(span, enduser.Conditions) {
+		return
+	}
+
+	var u *user
+	switch enduser.Type {
+	case "id":
+		u = p.idCapture(value)
+	case "role":
+		u = p.roleCapture(value)
+	case "scope":
+		u = p.scopeCapture(value)
+	case "authheader":
+		u = p.authHeaderCapture(enduser, value)
+	case "json":
+		u = p.jsonCapture(enduser, value)
+	case "urlencoded":
+		u = p.urlencodedCapture(enduser, value)
+	case "cookie":
+		u = p.cookieCapture(enduser, value)
+	default:
+		p.logger.Warn("Unknown enduser type, skipping", zap.String("type", enduser.Type), zap.String("key", enduser.AttributeKey))
```

Comment:
I think `type` is a very generic key. I think the idea of having this fields is to be able to filter by them so maybe we use something more specific? cc @davexroth